Use fake services to replicate avalanche target

arajkumar opened the pull request:
Currently we use k8s StatefulSet and native replica to create multiple Avalanche instances to generate series load to Prometheus. However this is less optimal and occupies real CPU & memory in benchmarking environment.

This fix puts fake k8s services infront of single avalanche Pod to mimic existence of multiple avalanche with less demanding resources.

Signed-off-by: Arunprasad Rajkumar <ar.arunprasad@gmail.com>

<!--
Thank you for contributing to timescale/tobs.
Before you submit this pull request we'd like to make sure you are aware of our technical requirements and best practices:

* https://github.com/prometheus-community/helm-charts/blob/main/CONTRIBUTING.md#technical-requirements
* https://helm.sh/docs/chart_best_practices/

Following our best practices right from the start will accelerate the review process and help get your pull request merged quicker.

Please make sure you test your changes before you push them.
Once the pull request is opened, GitHub Actions will run across your changes and do some initial checks and linting.
These checks run very quickly.
Please check the results.
If you are contributing to this repository for the first time, a maintainer will need to approve those checks to run.
They are automatically requested as reviewers and will approve the workflows or ask you for changes once they get to it.

We would like these checks to pass before we even continue reviewing your changes.
-->

<!-- markdownlint-disable-next-line first-line-heading -->
#### What this PR does / why we need it



#### Which issue this PR fixes

*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*

- fixes #

#### Special notes for your reviewer


#### Checklist

<!-- [Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.] -->
- [X] Chart Version bumped
- [X] [CLA signed](https://cla-assistant.io/timescale/helm-charts)

paulfantom reviewed (changes requested):
Since the idea here is to allow scraping more data while creating fake targets by scaling Service objects, I don't see why we need to switch from using STS to using static Pod. Such switch has negative impact on a scenario where avalanche is sending data via remote_write.

If we want to allow generating data via "fake targets", imho it would be easier to just switch from ServiceMonitor to PodMonitor and scale the latter.

arajkumar commented on charts/avalanche/values.yaml:
```diff
@@ -6,9 +6,17 @@ image:
+# Number of fake PodMonitor to be created to simulate fake Prometheus
+# targets. This will be useful to generate more series without scaling
+# Avalanche pods when using it as a Prometheus target.
+fakePrometheusTargetCount: 0
```
I see that this is not backward compatible, Since there is no ServiceMonitor, pulling this change would break the update. Let me fix it.

paulfantom commented:
We are using STS as we need stable network identifiers and STS is providing them out of the box. We don't want to use static pods as it is not a canonical way of managing any application in kubernetes and it should be used only in very small subset of specific cases.

paulfantom commented on charts/avalanche/templates/pod-monitor.yaml:
```diff
@@ -0,0 +1,22 @@
+{{ if not .Values.serviceMonitor.enabled }}
```
```suggestion
{{ if not .Values.podMonitor.enabled }}
```

paulfantom commented on charts/avalanche/values.yaml:
```diff
@@ -49,6 +52,12 @@ labels: {}
 serviceMonitor:
```
```suggestion
podMonitor:
```

paulfantom reviewed (changes requested):
Overall lgtm, it just need few changes on available configuration options.

arajkumar commented:
@paulfantom Adding multiple service monitor's not creating multiple targets, instead I created multiple services pointing to same pod which indeed creates multiple prometheus targets.

arajkumar commented:
@paulfantom CI failure is due the following line where the grep exits with 1 when only avalanche charts are modified.
```
changed=$(ct list-changed --target-branch ${{ github.event.repository.default_branch }} | grep promscale)
```

I fixed it in the latest commit, PTAL.

onprem commented on charts/avalanche/templates/svc-avalanche.yaml:
```diff
@@ -1,10 +1,9 @@
 kind: Service
 metadata:
-  name: {{ include "avalanche.fullname" . }}
+  name: {{ include "avalanche.fullname" . }}-{{ $n }}
```
We are using `serviceName: {{ include "avalanche.fullname" . }}` in the spec of the StatefulSet, this change will probably break the stable network IDs had previously. I think we can add another headless service with `name: {{ include "avalanche.fullname" . }}` to avoid breaking this behaviour. WDYT?

onprem replied:
We need to keep in mind that this headless service is excluded from the ServiceMonitor, so that we don't end up increasing the load data unintentionally. Maybe we can add an additional label to all scrapable services and add the same in `selector` section of ServiceMonitor.

onprem commented on charts/avalanche/values.yaml:
```diff
@@ -49,6 +52,11 @@ labels: {}
 serviceMonitor:
   enabled: false
 
+# Number of fake ServiceMonitor to be created to simulate fake Prometheus
```
nit:
 
```suggestion
# Number of Service objects to be created to simulate fake Prometheus
```
